h/static/scripts/controllers/share-widget-controller.js

```diff
@@ -0,0 +1,211 @@
+'use strict';
+
+const Controller = require('../base/controller');
+
+const TEMPLATE_SELECTOR = '.share-widget-template';
+const CONFIG_ATTR = 'share-widget-config';
+const TRIGGER_SELECTOR = `[${CONFIG_ATTR}]`;
+const WIDGET_SELECTOR = '.share-widget';
+const TARGET_HREF_ATTR = 'share-target-href';
+const TARGET_HREF_SELECTOR = `[${TARGET_HREF_ATTR}]`;
+const CLIPBOARD_INPUT_SELECTOR = 'input.share-widget-link__text';
+const PRIVATE_MSG_SELECTOR = '.share-widget-msg-private';
+const GROUP_MSG_SELECTOR = '.share-widget-msg-group';
+
+const ARROW_PADDING_RIGHT =  16;
+const ARROW_PADDING_BOTTOM =  5;
+
+
+// Travel up the nodes parents to find a matching
+// element with the share-widget configuration values
+const _getParentWithConfig = function(node) {
+  let configNode = null;
+  while ((node = node.parentNode)) {
+    if (node === document.body) {
+      break;
+    }
+    if (node.hasAttribute(CONFIG_ATTR)) {
+      configNode = node;
+      break;
+    }
+  }
+  return configNode;
+};
+
+const _getOffset = function (el) {
+  el = el.getBoundingClientRect();
+  return {
+    // adjust for top left of the document
+    left: el.left + window.scrollX,
+    top: el.top + window.scrollY,
+    width: el.width,
+    height: el.height,
+  };
+};
+
+
+class ShareWidget extends Controller {
+
+  constructor(element, options = {}) {
+    super(element, options);
+
+    // we are setting up a global listener for
+    // share widgets. It will only actually run once
+    // but having a widget at all on the page and all
+    // other share-widget components will be handled
+    // through psuedo event delegation.
+    this._setupGlobalWidget();
+  }
+
+  _setupGlobalWidget() {
+
+    if (!window.h || !window.h.shareWidget) {
+
+      window.h = Object.assign({}, window.h, {
+
+        shareWidget: {
```

Comment:
If `shareWidget` is a class rather than just a bag of properties I suggest we follow our conventions and make it an actual `class` at the top of the file using standard method syntax rather than `function` expressions.